pkg/tfshim/sdk-v2/provider2.go

```diff
@@ -798,3 +803,83 @@ func recoverDiagnostic(d tfprotov5.Diagnostic) diag.Diagnostic {
 	}
 	return dd
 }
+
+func normalizeBlockCollections(val cty.Value, res *schema.Resource) cty.Value {
+	// Full rules about block vs attr
+	//nolint:lll
+	// https://github.com/hashicorp/terraform-plugin-sdk/blob/1f499688ebd9420768f501d4ed622a51b2135ced/helper/schema/core_schema.go#L60
+	sch := res.CoreConfigSchema()
+	if !val.Type().IsObjectType() {
+		contract.Failf("normalizeBlockCollections: Expected object type, got %v", val.Type().GoString())
+	}
+	valMap := val.AsValueMap()
+
+	for fieldName := range sch.BlockTypes {
+		if !val.Type().HasAttribute(fieldName) {
+			continue
+		}
+		subVal := val.GetAttr(fieldName)
+		if subVal.IsNull() {
+			fieldType := val.Type().AttributeType(fieldName)
+			// Only lists and sets can be blocks and pass InternalValidate
+			// Ignore other types.
+			if fieldType.IsListType() {
+				glog.V(10).Info("normalizing list %s, %s", fieldName, fieldType.ElementType())
+				valMap[fieldName] = cty.ListValEmpty(fieldType.ElementType())
+			} else if fieldType.IsSetType() {
+				glog.V(10).Info("normalizing set %s, %s", fieldName, fieldType.ElementType())
+				valMap[fieldName] = cty.SetValEmpty(fieldType.ElementType())
+			}
+		} else {
+			subBlockSchema := res.SchemaMap()[fieldName]
+			if subBlockSchema == nil {
+				glog.V(5).Info("normalizeBlockCollections: Unexpected nil subBlockSchema for %s", fieldName)
+				continue
+			}
+			subBlockRes, ok := subBlockSchema.Elem.(*schema.Resource)
+			if !ok {
+				glog.V(5).Info("normalizeBlockCollections: Unexpected schema type %s", fieldName)
+				continue
+			}
+			glog.V(10).Info("normalizing block %s %s", fieldName, subVal.Type().GoString())
+			normalizedVal := normalizeSubBlock(subVal, subBlockRes)
+			valMap[fieldName] = normalizedVal
+		}
+	}
+	return cty.ObjectVal(valMap)
+}
+
+func normalizeSubBlock(val cty.Value, subBlockRes *schema.Resource) cty.Value {
+	if !val.IsKnown() {
+		// Blocks shouldn't be unknown, but if they are, we can't do anything with them.
+		return val
+	}
+	if val.Type().IsListType() {
+		newSlice := normalizeIterable(val, subBlockRes)
+		if len(newSlice) != 0 {
+			return cty.ListVal(newSlice)
+		}
+		return cty.ListValEmpty(val.Type().ElementType())
+	} else if val.Type().IsSetType() {
+		newSlice := normalizeIterable(val, subBlockRes)
+		if len(newSlice) != 0 {
+			return cty.SetVal(newSlice)
+		}
+		return cty.SetValEmpty(val.Type().ElementType())
+	}
+	return val
+}
+
+func normalizeIterable(blockVal cty.Value, blockRes *schema.Resource) []cty.Value {
+	contract.Assertf(blockVal.Type().IsListType() || blockVal.Type().IsSetType(),
+		"normalizeIterable: Expected list or set type, got %v", blockVal.Type().GoString())
+	if blockVal.IsNull() || !blockVal.IsKnown() {
```

Comment:
Are we certain about the  `!blockVal.IsKnown()` case? This seems like a bigger deal. Tested this well?

Reply:
I've opened https://github.com/pulumi/pulumi-terraform-bridge/issues/2032 to follow up with an investigation of unknown blocks, which is inexpressible in TF.

It should not happen but if it does we can't do anything about it in that function, so I've opted to keep it as is for now.